library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/net/BatchesToSegmentsMapper.kt

```diff
@@ -42,74 +43,108 @@ class BatchesToSegmentsMapper {
+            return null
+        }
 
-        @Suppress("UnsafeThirdPartyFunctionCall")
-        val stopTimestamp = records.last().getAsJsonPrimitive(TIMESTAMP_KEY).asLong
-        val hasFullSnapshotRecord = hasFullSnapshotRecord(records)
+        val startTimestamp = orderedRecords.firstOrNull()?.safeGetAsJsonObject()?.timestamp()
+        val stopTimestamp = orderedRecords.lastOrNull()?.safeGetAsJsonObject()?.timestamp()
+
+        if (startTimestamp == null || stopTimestamp == null) {
+            // this is just to avoid having kotlin warnings but the elements
+            // without timestamp property were already removed in the logic above
+            // TODO: RUMM-2397 Add the proper logs here once the sdkLogger will be added
+            return null
+        }
+
+        val hasFullSnapshotRecord = hasFullSnapshotRecord(orderedRecords)
         val segment = MobileSegment(
-            application = MobileSegment.Application(entry.key.applicationId),
-            session = MobileSegment.Session(entry.key.sessionId),
-            view = MobileSegment.View(entry.key.viewId),
+            application = MobileSegment.Application(rumContext.applicationId),
+            session = MobileSegment.Session(rumContext.sessionId),
+            view = MobileSegment.View(rumContext.viewId),
             start = startTimestamp,
             end = stopTimestamp,
-            recordsCount = records.size.toLong(),
+            recordsCount = orderedRecords.size().toLong(),
             // TODO: RUMM-2518 Find a way or alternative to provide a reliable indexInView
             indexInView = null,
             hasFullSnapshot = hasFullSnapshotRecord,
             source = MobileSegment.Source.ANDROID,
             records = emptyList()
         )
-        val segmentAsJsonObject = segment.toJson().asJsonObject
-        segmentAsJsonObject.add(RECORDS_KEY, entry.value)
+        val segmentAsJsonObject = segment.toJson().safeGetAsJsonObject() ?: return null
+        segmentAsJsonObject.add(RECORDS_KEY, orderedRecords)
         return Pair(segment, segmentAsJsonObject)
     }
 
-    private fun hasFullSnapshotRecord(records: List<JsonObject>) =
+    private fun hasFullSnapshotRecord(records: JsonArray) =
         records.firstOrNull {
-            it.getAsJsonPrimitive(RECORD_TYPE_KEY).asLong == FULL_SNAPSHOT_RECORD_TYPE
+            it.asJsonObject.getAsJsonPrimitive(RECORD_TYPE_KEY)?.asLong == FULL_SNAPSHOT_RECORD_TYPE
         } != null
 
+    private fun JsonObject.records(): JsonArray? {
+        return get(EnrichedRecord.RECORDS_KEY)?.asJsonArray
+    }
+
+    private fun JsonObject.timestamp(): Long? {
+        return getAsJsonPrimitive(TIMESTAMP_KEY)?.asLong
+    }
+
+    private fun JsonObject.rumContext(): SessionReplayRumContext? {
+        val applicationId = get(EnrichedRecord.APPLICATION_ID_KEY)?.asString
+        val sessionId = get(EnrichedRecord.SESSION_ID_KEY)?.asString
+        val viewId = get(EnrichedRecord.VIEW_ID_KEY)?.asString
```

Comment:
these methods can still throw json related exceptions and they are not covered by try-catch